### PR TITLE
30/31 Creates an account from a submission + sends confirmation email

### PR DESCRIPTION
**Pull request by zerolab**

### CODE REVIEW

File: opentech/apply/funds/blocks.py [hunk 1]
```diff
@@ -140,4 +140,15 @@ class ValueBlock(MustIncludeFieldBlock):
     widget = forms.NumberInput
 
 
+class EmailBlock(MustIncludeFieldBlock):
+    name = 'email'
+    description = 'The applicant email address'
+    widget = forms.EmailInput
+
```
Comment (todd-dembrey) [inline opentech/apply/funds/blocks.py]:
```
    class Meta:
        icon = 'email'
```

File: opentech/apply/funds/blocks.py [hunk 2]
```diff
@@ -140,4 +140,15 @@ class ValueBlock(MustIncludeFieldBlock):
...
+
+class FullNameBlock(MustIncludeFieldBlock):
+    name = 'full_name'
+    description = 'Full name'
```
Comment (todd-dembrey) [inline opentech/apply/funds/blocks.py]: Maybe...?
```
    class Meta:
        icon = 'user'
```

File: opentech/apply/funds/models.py [hunk 3]
```diff
@@ -74,6 +77,18 @@ def next_deadline(self):
...
+                FieldRowPanel([
+                    FieldPanel('from_address', classname="col6"),
+                    FieldPanel('to_address', classname="col6"),
+                ]),
+                FieldPanel('subject'),
+                FieldPanel('confirmation_text_extra'),
+            ],
+            heading="Confirmation email",
+            classname="collapsible collapsed"
+        ),
```
Comment (todd-dembrey) [inline opentech/apply/funds/models.py]: Maybe a new tab for this? "email settings" 

Not essential, but my preference.

http://docs.wagtail.io/en/v1.13.1/advanced_topics/customisation/page_editing_interface.html#customising-the-tabbed-interface

File: opentech/apply/funds/models.py [hunk 4]
```diff
@@ -143,12 +158,60 @@ def process_form_submission(self, form):
                 response = cleaned_data.pop(field.id)
                 cleaned_data[field.block.name] = response
 
+        if cleaned_data.get('email'):
```
Comment (todd-dembrey) [inline opentech/apply/funds/models.py]: If we are here then the required field "email" will exist so we don't need to check if an email is provided. 

I think the rest of this section should be a new method `create_user`.

File: opentech/apply/funds/models.py [hunk 5]
```diff
@@ -143,12 +158,60 @@ def process_form_submission(self, form):
...
+                user_data = {
+                    'username': full_name if full_name else email,
+                    'email': email,
+                    'is_active': False,  # User needs to activate account
+                }
+
+                if full_name and hasattr(User, 'full_name'):
+                    user_data['full_name'] = full_name
+
+                user = User.objects.create_user(**user_data)
```
Comment (todd-dembrey) [inline opentech/apply/funds/models.py]: This should be the only part wrapped in the try except. 

```
try:
    user = User.objects.create_user(**user_data)
except IntegrityError:
    user = User.objects.get(email=email)
else:
    # send the activation
```

I think I would be happy to move the `send_activation_email` as part of our `create_user` method on the custom User.

Can you think of a reason not to?

File: opentech/apply/funds/models.py [hunk 6]
```diff
@@ -143,12 +158,60 @@ def process_form_submission(self, form):
...
+
+                if user:
+                    # TODO: move to method on User?
+                    from opentech.apply.users.utils import send_activation_email
+                    send_activation_email(user)
+
+            except IntegrityError:
+                pass
+
+            self.send_confirmation_email(form, cleaned_data)
```
Comment (todd-dembrey) [inline opentech/apply/funds/models.py]: pass in `user` ?

File: opentech/apply/funds/models.py [hunk 7]
```diff
@@ -143,12 +158,60 @@ def process_form_submission(self, form):
...
+    def send_confirmation_email(self, form, cleaned_data):
+        if not cleaned_data.get('email'):
+            return
+
+        from django.template.loader import render_to_string
+        from wagtail.wagtailadmin.utils import send_mail
+
+        email = cleaned_data.get('email')
+        fund = self.get_parent().specific
+        context = {
```
Comment (todd-dembrey) [inline opentech/apply/funds/models.py]: `to_address` not used?

File: opentech/apply/funds/models.py [hunk 8]
```diff
@@ -143,12 +158,60 @@ def process_form_submission(self, form):
...
+
+            self.send_confirmation_email(form, cleaned_data)
+
         return self.get_submission_class().objects.create(
             form_data=cleaned_data,
             page=self.get_parent(),
             round=self,
         )
 
+    def send_confirmation_email(self, form, cleaned_data):
```
Comment (todd-dembrey) [inline opentech/apply/funds/models.py]: might be worth making this generic as we will need to apply this to labs. Can do this at a later date.

File: opentech/apply/users/views.py [hunk 9]
```diff
@@ -20,8 +25,62 @@ def account(request):
...
+
+class ActivationView(TemplateView):
     """
-    Generic, empty view for the OAuth associations.
+    Inspired by https://github.com/ubernostrum/django-registration
     """
 
-    return TemplateResponse(request, 'users/oauth.html', {})
+    def get(self, request, *args, **kwargs):
+        user = self.activate(*args, **kwargs)
```
Comment (todd-dembrey) [inline opentech/apply/users/views.py]: Not too happy with the way this reads. Would prefer something like below which means you don't need to deconstruct the class to understand.

```
user = self.get_user(uidb64)
if self.valid(user, token):
    self.activate(user)
    login(user)
    return redirect()
return redirect('invalid')
```

File: opentech/apply/funds/models.py (hunk 5 above)
Comment (zerolab) [reply opentech/apply/funds/models.py]: the only reason against it is that the only use case for this is when the user is created through submission

Comment (todd-dembrey) [reply opentech/apply/funds/models.py]: I'm not sure i understand. You should still tighten up the try/except even if the except doesnt set the user.

File: opentech/apply/funds/models.py (hunk 7 above)
Comment (zerolab) [reply opentech/apply/funds/models.py]: see `(email,)` in `send_mail()`

File: opentech/apply/users/views.py (hunk 9 above)
Comment (zerolab) [reply opentech/apply/users/views.py]: 👍

File: opentech/apply/funds/models.py (hunk 7 above)
Comment (todd-dembrey) [reply opentech/apply/funds/models.py]: I'm talking about `fund.to_address` I'm assuming that is for cc/bcc?

Comment (zerolab) [reply opentech/apply/funds/models.py]: yes. that is handled by `AbstractEmailForm`

File: opentech/apply/funds/blocks.py (hunk 1 above)
Comment (zerolab) [reply opentech/apply/funds/blocks.py]: done

File: opentech/apply/funds/blocks.py (hunk 2 above)
Comment (zerolab) [reply opentech/apply/funds/blocks.py]: done

**Comment by zerolab** [comment]
Superseded by #76

**Pull request closed without merging**